### DISCUSSION TITLE
This can scale to relatively big programs?

### DISCUSSION THREAD

**Discussion by mamcx** [egraphs-good/egg#273] [General]
I'm exploring using this library for a full language that embeds a query (think PSQL), so I have 2 distinct use cases:

- The user sends a `SQL` query that I compile to my AST, here I apply a query optimizer
- The user write a full app backend, like the one you make in C# + LINQ, so here also is a compiler optimizer, so this is a relatively "big" program with dozens of files or more

So, is desirable that I can resolve very fast the first case, and have a good compiler speed in the second.

I'm ok having restrictions, the lang is in developing so I'm free to change it so it become easy to optimize.

In specific, I wanna optimize things like:

- Basic query optimizations like in https://rustmagazine.org/issue-2/write-a-sql-optimizer-using-egg/
-  For loop rewrites of the one you see in Rust + Iterators
- Array operations like `[1, 2] + 1`
- Basic compiler optimization like constant folding, constant rewrite, etc

**Comment by mwillsey** [comment]
It's possible to scale to larger programs, but it will take work. This
technique is very sensitive to the rewrites that you are using; for
example, associativity and commutativity are known to grow the size of the
e-graph rapidly. You will also probably have to do things like use a rule
scheduler, or manage e-graph size by extracting and re-optimizing a term.
